crates/ruma-state-res/tests/it/resolve.rs

```diff
@@ -0,0 +1,317 @@
+    }
+}
+
+impl Eq for ResolvedState {}
+
+impl Ord for ResolvedState {
+    fn cmp(&self, other: &Self) -> Ordering {
+        Ordering::Equal
+            .then(self.kind.cmp(&other.kind))
+            .then(self.state_key.cmp(&other.state_key))
+            .then(self.event_id.cmp(&other.event_id))
+    }
+}
+
+impl PartialOrd for ResolvedState {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// Results of a test.
+struct TestResults {
+    /// The resolved state of the room.
+    resolved_state: BTreeSet<ResolvedState>,
+}
+
+/// Test a list of JSON files containing a list of PDUs and return the results.
+///
+/// State resolution is run both atomically for all PDUs and in batches of PDUs by file.
+fn run<P>(paths: &[P]) -> Result<TestResults, Box<dyn Error>>
+where
+    P: AsRef<Path>,
+{
+    // Run `cargo test -- --show-output` to view traces, set `RUST_LOG` to control filtering.
+    let _subscriber = tracing_subscriber::fmt()
+        .with_env_filter(EnvFilter::from_default_env())
+        .with_test_writer()
+        .finish()
+        .set_default();
+
+    let fixtures_path = PathBuf::from("tests/it/fixtures");
+
+    // `serde_json::from_reader` is disallowed for normal code, but this is just for testing.
+    #[allow(clippy::disallowed_methods)]
+    let pdu_batches = paths
+        .iter()
+        .map(|x| serde_json::from_reader(File::open(fixtures_path.join(x))?).map_err(Into::into))
+        .collect::<Result<Vec<Vec<Pdu>>, Box<dyn Error>>>()?;
+
+    let room_version = {
+        let first_pdu = pdu_batches
+            .first()
+            .ok_or("there should be at least one file of PDUs")?
+            .first()
+            .ok_or("there should be at least one PDU in the first file")?;
+
+        if first_pdu.kind != TimelineEventType::RoomCreate {
+            return Err("the first PDU in the first file should be an m.room.create event".into());
+        }
+
+        serde_json::from_str::<ExtractRoomVersion>(first_pdu.content.get())
+            .map_err(|_| "the m.room.create PDU's content is invalid")?
+            .room_version
+    };
+
+    // Resolve PDUs in batches by file
+    let mut pdus_by_id = HashMap::new();
+    let mut batched_resolved_state = None;
+    for pdus in &pdu_batches {
+        batched_resolved_state =
+            Some(resolve_batch(&room_version, pdus, &mut pdus_by_id, &mut batched_resolved_state)?);
+    }
+    let batched_resolved_state =
+        batched_resolved_state.ok_or("state resolution should have run at least once")?;
+
+    // Resolve all PDUs in a single step
+    let atomic_resolved_state = resolve_batch(
+        &room_version,
+        pdu_batches.iter().flat_map(|x| x.iter()),
+        &mut HashMap::new(),
+        &mut None,
+    )?;
+
+    if batched_resolved_state != atomic_resolved_state {
+        return Err("batched and atomic resolved states should match".into());
+    }
```

Comment:
The differences could be pretty large. I've added a commit called "state-res: wip: show atomic/batched diff" that will print out a diff if they don't match; let me know what you think of that idea.

Reply:
We don't need to see the `content` though, that just makes the diff bigger. We could also use `similar_asserts` directly.

Reply:
By "the values" I meant the `StateMap`s themselves, because there are currently 6-8 events in them. Regardless of whether `content` is present, trying to compare these in non-pretty-printed `Debug` format would be pretty painful, which is why I don't think `assert_eq` is a good choice here. I think using the same format in the diff as in the snapshots also makes it clearer what's going on.

> We could also use `similar_asserts` directly.

We could. I didn't know that crate existed and `similar` was already in our dependency tree and I wouldn't have been sure if adding more dependencies would have been well-received.

Reply:
The current code serializes to JSON and diffs it, `similar_asserts` doesn't seem to have an easier way to do that (it serializes to a custom format), so I think we should stick with the current version so it's easier to see the relationship to the snapshot files.